### PR TITLE
docs: convert docs to use mkdocs

### LINKED ISSUE #1708: Address various issues in mkdocs for Python
# Description

We want to use mkdocs instead, but need to address some issues in the draft.

## Issue 1: cross reference aren't working

```
WARNING -  mkdocs_autorefs: usage/index.md: Could not find cross-reference target '[DeltaTable]'
WARNING -  mkdocs_autorefs: usage/examining-table.md: Could not find cross-reference target '[DeltaTable.metadata()]'
WARNING -  mkdocs_autorefs: usage/examining-table.md: Could not find cross-reference target '[DeltaTable.schema]'
WARNING -  mkdocs_autorefs: usage/examining-table.md: Could not find cross-reference target '[deltalake.schema.Schema.from_json()]'
WARNING -  mkdocs_autorefs: usage/examining-table.md: Could not find cross-reference target '[deltalake.schema.Schema.to_pyarrow()]'
```

## Issue 2: API Reference table of contents is too flat for schema types

image

One you get the schema module, it stops nesting the methods

## Issue 3: API reference contains random dunder methods for schema types

It shows all the `__eq__()`, `__ge__()`, etc. methods for `PrimitiveType`, for example.

## Issue 4: Old sphinx directives in docstrings need to be replaced

We use Sphinx directives like `.. deprecated::` in the docstrings. They need to be replaced with the mkdocs alternative.

## Issue 5: Examples don't render properly

Screenshot 2023-10-07 at 2 22 40 PM

## Issue 6: docstring fields aren't being parsed

Things like `:param` and `:returns` are just shown raw.

# TODO

* [ ] Fix cross references in mkdocs
* [ ] Fix API reference table of contents hierarchy in mkdocs
* [ ] Hide dunder methods in mkdocs API reference
* [ ] Replace old sphinx directives with mkdocs equivalents in docstrings
* [ ] Make docstring examples display properly in mkdocs
* [ ] Format docstrings so mkdocs can understand parameters and return types.

### PR DESCRIPTION
**Pull request by r3stl355**
# Description
Completed the outstanding tasks in #1708

Also changed theme from readthedocs to mkdocs - both are built-in but latter looks sleeker

# Related Issue(s)
closes #1708

### CODE REVIEW

**Comment by rtyler** [comment]
@r3stl355 thanks for the change, there's some formatting that should be applied (e.g. `cargo fmt && cd python && make format`) and conflicts with `main` resolved

**Comment by r3stl355** [comment]
ok, missed those, will fix and resolve

**Comment by r3stl355** [comment]
revolved all, also hosted here for better visualization: https://delta-rs-docs.pages.dev/

**Comment by r3stl355** [comment]
this merge with main for conflicts caught me off guard a bit, let me know if you want me to squash all these little commits

**Comment by r3stl355** [comment]
sphinx build is failing due to warnings with indentation of multi-line parameter descriptions in docstrings. If I remove the indentation then mkdocs starts throwing warnings, so it's one or another I guess

**Comment by MrPowers** [comment]
I am so excited about this.  Great work @r3stl355!!

File: mkdocs.yml [hunk 1]
```diff
@@ -14,22 +16,30 @@ nav:
...
-          - https://pandas.pydata.org/docs/objects.inv
+  - autorefs
+  - mkdocstrings:
+      handlers:
+        python:
+          path: [../python]
+          options:
+            # docstring_style: sphinx
+            docstring_section_style: table
+            ocstring_section_style: litabst
```
Comment (roeap) [inline mkdocs.yml:27]:
```suggestion
            docstring_section_style: litabst
```

**Review by roeap — COMMENTED** [review]
excited to see this land!

**Comment by r3stl355** [comment]
With frequent changes to main this PR is getting out of date pretty quick. I think this PR is needed as a foundation for building better docs (e.g. docs referring to both Python and Rust from a single place). Let me know if I can take an alternative approach - e.g. split this into a smaller set of PRs

**Review by rtyler — APPROVED** [review]

**Comment by r3stl355** [comment]
thank you for merging this @rtyler, I'll start working on the next steps, will raise an issue to track the progress

**Pull request merged**